src/tainting/Dataflow_tainting.ml

```diff
@@ -423,45 +423,85 @@ and eval_label_args ~labels args =
+  (* We cannot check whether we satisfy the `requires` here.
+     This is because this sink may be inside of a function, meaning that
+     argument taint can reach it, which can only be instantiated at the
+     point where we call the function.
+     So we record the `requires` within the taint finding, and evaluate
+     the formula later, when we extract the PMs
+  *)
   let sink_pm, ts = T.pm_of_trace sink in
-  let req = eval_label_requires ~labels ts.sink_requires in
-  (* TODO: With taint labels it's less clear what is "the source",
-     * in fact, there could be many sources, each one providing a
-     * different label. Here these would be reported as different
-     * findings... We would need to have multiple sources in `SrcToSink`!
-     * And `ArgtoSink` needs to carry the other taint that reaches the
-     * sink besides the argument. *)
-  taints |> Taints.elements
-  |> List.filter_map (fun (taint : T.taint) ->
-         let tokens = List.rev taint.tokens in
-         match taint.orig with
-         | Arg i ->
-             (* We need to check the label and unifiability requirements
-                at the call site. *)
-             Some (T.ArgToSink (i, tokens, sink))
-         | Src source ->
-             if req then
-               let src_pm, _ = T.pm_of_trace source.call_trace in
-               let* merged_env =
-                 merge_source_sink_mvars env sink_pm.PM.env src_pm.PM.env
-               in
-               Some (T.SrcToSink { source; tokens; sink; merged_env })
-             else None)
+  let taints_and_bindings =
+    Taints.elements taints
+    |> Common.map (fun t ->
+           let bindings =
+             match t.T.orig with
+             | T.Arg _ -> []
+             | Src source ->
+                 let src_pm, _ = T.pm_of_trace source.call_trace in
+                 src_pm.PM.env
+           in
+           ({ t with tokens = List.rev t.tokens }, bindings))
+  in
+  (* If `unify_mvars` is set, then we will just do the previous behavior,
+     and emit a finding for every single source coming into the sink.
+     This will mean we don't regress on `taint_unify_mvars: true` rules.
+
+     This is problematic because there may be many sources, all of which do not
+     unify with each other, but which unify with the sink.
+     If we did as below and unified them all with each other, we would sometimes
+     produce no findings when we should.
+  *)
+  if env.config.unify_mvars then
+    taints_and_bindings
+    |> List.filter_map (fun (t, bindings) ->
+           let* merged_env =
+             merge_source_sink_mvars env sink_pm.PM.env bindings
+           in
+           Some
+             (T.ToSink
+                {
+                  taints_with_precondition = ([ t ], R.get_sink_requires ts);
+                  sink;
+                  merged_env;
+                }))
+  else
+    match
+      taints_and_bindings |> List.concat_map snd
+      |> merge_source_sink_mvars env sink_pm.PM.env
```

Comment:
If source A binds `$X` and source B binds `$X`, and the sink does NOT bind `$X`, then I believe this will return a list of bindings where `$X` will occur twice possibly bound to different values.

Reply:
oops, good catch. should be fixed now.